### PR TITLE
Fix `StringSegment.Buffer` can be null

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Addresses: https://github.com/dotnet/runtime/pull/57395#issuecomment-902859880

- Made `Buffer` nullable
- `Equals(text)` returns `true` if both `Buffer` and `text` are null

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs
Comment (gfoidl) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:105]: I'd prefer
```suggestion
                Debug.Assert(Buffer is not null);
                return Buffer[Offset + index];
```
because if in the ctor a bug get introduced that doesn't set the correct Length when buffer = null, then it will blow up here.

**Comment by drieseng** [comment]
Perhaps also add a test where **null** is passed to .ctor(String), and assert that:
* **Buffer** returns **null** (and ...)
* **Offset** return **zero**
* **Length** returns **zero**
* **HasValue** returns **false**
* **Value** returns **null**
* ...

**Comment by maxkoshevoi** [comment]
There is one already: https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs#L164
Nullable is not enabled for tests, so didn't notice it when doing https://github.com/dotnet/runtime/pull/57395

**Comment by drieseng** [comment]
@maxkoshevoi Missed that one, thx.

**Review by JamesNK — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:262]: If we are going to throw NullArg here, the `text` parameter needs to be updated to say `string text`, so callers know they shouldn't pass `null` in here.

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:262]: `text` cannot be non-nullable in `Equals(string? text)` (by interface contract): https://github.com/dotnet/runtime/pull/57395#discussion_r688975416

So it might be strange to make it non-nullable for `Equals([NotNullWhen(true)] string? text, StringComparison comparisonType)`: https://github.com/dotnet/runtime/pull/57395#discussion_r688974246

This was discussed in the comments above.

I can make `text` non-nullable, but only in `Equals([NotNullWhen(true)] string? text, StringComparison comparisonType)`

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:262]: I don't understand why we'd throw here.  It's equals or it's not... what necessitate making it yes,  no,  or throw?

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:262]: I think, returning `false` from `new StringSegment(null).Equals((string?)null);` would be even less obvious.

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:262]: > No idea why it doesn't return true though.

Didn't you just change it to return false a few days ago and before that it was throwing?

We should make it return whatever value makes sense, given the combination of the current instance and the provided value.  Throwing doesn't make sense,  at least not to me, given that Equals in general allows null arguments.

Comment (JamesNK) [reply src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:262]: StringSegment has an implicit conversion from string. What about logic equivalent to:

```cs
public bool Equals([NotNullWhen(true)] string? text, StringComparison comparisonType)
{
    if (text == null)
    {
        StringSegment other = text;
        return Equals(other);
    }

    /// ...
}
```

You should also update the implicit conversion to allow a nullable string (~line 316):

```diff
- public static implicit operator StringSegment(string value) => new StringSegment(value);
+ public static implicit operator StringSegment(string? value) => new StringSegment(value);
```

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:262]: > What about logic equivalent to

This is what I meant by "Presumably segment.Equals(string) should behave the same as segment.Equals(new StringSegment(string))...?"

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs]: > Didn't you just change it to return false a few days ago and before that it was throwing?

Yep. I thought `new StringSegment(null)` was illegal. In this PR I didn't want to change behavior again without your blessing, so just reverted Equals implementation to what was originally there.

> Presumably segment.Equals(string) should behave the same as segment.Equals(new StringSegment(string))...?

Makes sense. Updated the implementation and tests so that Equals never throws.

> You should also update the implicit conversion to allow a nullable string (~line 316):

Thank you, missed that one.

File: src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs:1527]: Why is this being changed?

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs:1527]: Tried to fix `StringSegment_CompareEqual_Globalized` test. This didn't work, so I'll revert it.
How does `new StringSegment()` compile? `StringSegment` doesn't have parameterless constructor, does it?

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs:1527]: > How does new StringSegment() compile?

It is a struct. Structs always have a parameterless constructor.

File: src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs:465]: Why is `{ null, StringComparison.OrdinalIgnoreCase, false },` being removed?

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs:465]: That's strange. I looked through the commits, and I've never removed it.
Added it back.

**Comment by JamesNK** [comment]
FYI this PR is blocking the runtime -> aspnetcore merge. If it is in a good state then it would be great to get it merged and unblock updates.

**Review by eerhardt — APPROVED** [review]

**Comment by stephentoub** [comment]
Thanks for fixing it.

**Pull request merged**
